File: src/utils/taxonomy.js
import promptsData from '../../prompts.json';
import { formatTaxonomyLabel } from '../constants';

/**
 * Résolution et affichage de la taxonomie — source unique côté frontend.
 *
 * ⚠️ POURQUOI DEUX NORMALISATIONS DISTINCTES (bug trouvé le 2026-08-16)
 * L'ancienne implémentation normalisait un chemin complet avec le même helper que les libellés,
 * lequel supprime TOUS les caractères non alphanumériques — points compris. Conséquence :
 *   normalize("Guitare Electrique")  === "guitareelectrique"
 *   normalize("guitare.electrique")  === "guitareelectrique"   ← MÊME CLÉ
 * La feuille `etui_housse.Etui_Rigide.Guitare Electrique` (un étui) entrait donc en collision avec
 * la branche `guitare.electrique` (l'instrument). Les chemins complets étant testés en premier, tout
 * étui classé par son nom de feuille était résolu comme une vraie guitare électrique — pas seulement
 * mal affiché : mal filtré et mal compté. Le séparateur de chemin est désormais préservé
 * (`normalizePath`), il ne peut plus se confondre avec du texte de libellé.
 */

// Normalise UN segment : minuscules, sans accents, alphanumérique uniquement.
export const normalizeSegment = (str) => {
  if (!str) return '';
  return String(str)
    .normalize('NFD').replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]/g, '');
};

// Normalise un CHEMIN en préservant le séparateur '.', qui porte du sens.
export const normalizePath = (str) =>
  String(str || '').split('.').map(normalizeSegment).filter(Boolean).join('.');

const MASTER_TAXONOMY = promptsData.taxonomy_master || {};

// Clés courtes et branche `etui_housse` exclues de la recherche floue : voir le commentaire
// historique de findPathFuzzy (collision "cordes guitare" ⊃ "sg", feuilles d'étuis qui reprennent
// mot pour mot les noms d'instruments).
const FUZZY_MIN_KEY_LENGTH = 5;
const FUZZY_EXCLUDED_BRANCHES = ['etui_housse'];

// Branche des contenants, dont CERTAINES feuilles portent des noms génériques d'instruments
// ("Guitare Acoustique", "Guitare Electrique") tout en désignant des étuis. Un tel nom nu n'est
// jamais résolu : il est bien plus probable qu'il désigne l'instrument lui-même.
// ⚠️ Restreint le 2026-08-16 après l'audit en production : la règle bloquait auparavant TOUTE
// feuille de la branche, y compris des termes d'étui purs sans ambiguïté ("Housse_Souple",
// "Gigbag Standard", "ATA Road Case") — ≈19 annonces correctement classées rendues invisibles.
// Miroir de `backend/taxonomy.py::CONTAINER_BRANCHES`.
const CONTAINER_BRANCHES = ['etui_housse'];

// Segments minimum d'un chemin partiel accepté pour une réparation par suffixe.
const MIN_PARTIAL_PATH_SEGMENTS = 2;

/**
 * Index de la taxonomie :
 *  - `byPath`  : chemin normalisé (points conservés) → tableau de segments
 *  - `byLeaf`  : nom de feuille/branche normalisé → TOUS les chemins qui portent ce nom.
 *                Un tableau, pas une valeur unique : plusieurs branches partagent des homonymes
 *                (`Guitare Acoustique` est à la fois un étui et un instrument, `solid_body` existe
 *                sous `electrique` ET sous `basse`). L'ancienne map écrasait silencieusement les
 *                doublons — la dernière branche parcourue gagnait, sans que rien ne le signale.
 *  - `nodes`   : catalogue plat pour l'autocomplétion.
 */
const buildIndex = () => {
  const byPath = {};
  const byLeaf = {};
  const nodes = [];

  const register = (segments) => {
    const pathKey = normalizePath(segments.join('.'));
    byPath[pathKey] = segments;

    // Deux clés pour un même nœud : son nom de feuille, ET son chemin complet privé de ses points.
    // C'est le cœur du bug des étuis : "Guitare Electrique" est une feuille de `etui_housse`, mais
    // aussi la lecture naturelle du CHEMIN `guitare.electrique`. Les enregistrer sous la même clé
    // rend le nom nu détectable comme ambigu, quel que soit le sens voulu par l'IA — sans quoi on
    // ne fait que déplacer l'erreur (un étui devenait une guitare ; l'inverse arriverait ensuite).
    const keys = new Set([
      normalizeSegment(segments[segments.length - 1]),
      normalizePath(segments.join('.')).replace(/\./g, '')
    ]);
    keys.forEach(key => {
      if (!key) return;
      if (!byLeaf[key]) byLeaf[key] = [];
      byLeaf[key].push(segments);
    });

    nodes.push({
      path: segments.join('.'),
      label: formatTaxonomyLabel(segments[segments.length - 1]),
      breadcrumb: segments.slice(0, -1).map(formatTaxonomyLabel).join(' › '),
      depth: segments.length,
      labelNormalized: segments[segments.length - 1]
        .normalize('NFD').replace(/[\u0300-\u036f]/g, '')
        .toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim()
    });
  };

  const traverse = (node, currentPath) => {
    if (Array.isArray(node)) {
      node.forEach(item => register([...currentPath, item]));
    } else if (typeof node === 'object' && node !== null) {
      Object.keys(node).forEach(key => {
        const segments = [...currentPath, key];
        register(segments);
        traverse(node[key], segments);
      });
    }
  };

  traverse(MASTER_TAXONOMY, []);
  return { byPath, byLeaf, nodes };
};

export const TAXONOMY_INDEX = buildIndex();
export const TAXONOMY_NODES = TAXONOMY_INDEX.nodes;

// Racines désignant des objets (guitare, amplificateur…), hors contenants — déduites de la
// taxonomie plutôt que codées en dur.
const INSTRUMENT_ROOTS = [...new Set(
  Object.values(TAXONOMY_INDEX.byPath)
    .map(segments => segments[0])
    .filter(root => !CONTAINER_BRANCHES.includes(root))
    .map(normalizeSegment)
)];

/**
 * Étend un chemin INCOMPLET vers l'unique chemin canonique qui s'y termine.
 * L'IA produit régulièrement des chemins presque bons : racine oubliée
 * (`electrique.solid_body.Double_Cut.SG`), niveau intermédiaire sauté, ou racine obsolète
 * (`accessoire_etui.protection.…`). Tous désignent une seule catégorie sans ambiguïté.
 * Miroir de `backend/taxonomy.py::_repair_partial_path`.
 */
const repairPartialPath = (classification) => {
  let segments = normalizePath(classification).split('.').filter(Boolean);
  const canonicalPaths = Object.values(TAXONOMY_INDEX.byPath);

  while (segments.length >= MIN_PARTIAL_PATH_SEGMENTS) {
    const suffix = `.${segments.join('.')}`;
    const matches = [...new Set(
      canonicalPaths
        .filter(path => normalizePath(path.join('.')).endsWith(suffix))
        .map(path => path.join('.'))
    )];
    if (matches.length === 1) return matches[0].split('.');
    if (matches.length > 1) return null; // Plusieurs cibles : on ne devine pas.
    segments = segments.slice(1);        // Racine inventée/obsolète : on la retire.
  }
  return null;
};

// Recherche floue : le chemin dont la clé textuelle est la plus longue (donc la plus spécifique)
// parmi celles contenues dans le texte. Dernier recours seulement — voir les exclusions ci-dessus.
const findPathFuzzy = (normalizedText) => {
  let bestKey = null;
  let bestPath = null;
  for (const [leafKey, paths] of Object.entries(TAXONOMY_INDEX.byLeaf)) {
    if (leafKey.length < FUZZY_MIN_KEY_LENGTH) continue;
    // Un nom ambigu (plusieurs nœuds distincts) ne peut pas être tranché par une recherche floue.
    const distinctPaths = [...new Set(paths.map(p => p.join('.')))];
    if (distinctPaths.length > 1) continue;
    const path = paths[0];
    if (path.length < 2) continue;
    if (FUZZY_EXCLUDED_BRANCHES.includes(path[0])) continue;
    if (normalizedText.includes(leafKey) && (!bestKey || leafKey.length > bestKey.length)) {
      bestKey = leafKey;
      bestPath = path;
    }
  }
  return bestPath;
};

/**
 * Résout la valeur `classification` renvoyée par l'IA en un chemin canonique.
 * Retourne `{ segments, ambiguous }` — `segments: null` si non résolue.
 * `ambiguous: true` signale un nom de feuille porté par plusieurs branches (ex: "Guitare
 * Acoustique") : impossible de trancher sans le chemin complet, donc volontairement NON résolu
 * plutôt que rattaché à une branche au hasard.
 */
export const resolveClassification = (classification) => {
  if (!classification) return { segments: null, ambiguous: false };

  // 1. Chemin complet exact (le format désormais exigé de l'IA)
  const asPath = normalizePath(classification);
  if (TAXONOMY_INDEX.byPath[asPath]) {
    return { segments: TAXONOMY_INDEX.byPath[asPath], ambiguous: false };
  }

  // 2. Nom de feuille seul (format historique, encore massivement présent en base).
  // Les candidats sont dédoublonnés : un même nœud est indexé sous plusieurs clés.
  const asLeaf = normalizeSegment(classification);
  const candidates = TAXONOMY_INDEX.byLeaf[asLeaf] || [];
  const distinct = [...new Set(candidates.map(c => c.join('.')))];
  if (distinct.length === 1) {
    const segments = distinct[0].split('.');
    // Feuille d'un contenant : refusée seulement si son nom commence par un nom de famille
    // d'instrument. Les termes d'étui purs passent normalement.
    if (CONTAINER_BRANCHES.includes(segments[0])) {
      const leafKey = normalizeSegment(segments[segments.length - 1]);
      if (INSTRUMENT_ROOTS.some(root => leafKey.startsWith(root))) {
        return { segments: null, ambiguous: true };
      }
    }
    return { segments, ambiguous: false };
  }
  if (distinct.length > 1) return { segments: null, ambiguous: true };

  // Chemin incomplet (racine oubliée, niveau sauté, racine obsolète) : extensible sans ambiguïté.
  const repaired = repairPartialPath(classification);
  if (repaired) return { segments: repaired, ambiguous: false };

  // 3. Dernier recours : recherche floue sur du texte libre (ex: "Fender Stratocaster")
  const fuzzy = findPathFuzzy(asLeaf);
  return { segments: fuzzy || null, ambiguous: false };
};

// Chemin en dot-notation (ex: "guitare.acoustique_acier.formes_standard.Parlor"), ou null.
export const resolveClassificationPath = (classification) => {
  const { segments } = resolveClassification(classification);
  return segments ? segments.join('.') : null;
};

/**
 * Libellé lisible destiné à l'UI.
 * La branche racine est préfixée quand elle n'est PAS `guitare` — sans ça, un étui s'affiche
 * "Guitare Electrique" et se lit comme une guitare (symptôme signalé par l'utilisateur), et un
 * ampli "Combo Lampes" perd son contexte. Pour une guitare, la feuille seule suffit et reste
 * courte ("Parlor Standard"), ce qui préserve l'affichage actuel des cartes.
 * Une valeur non résolue est retournée telle quelle mais formatée — jamais de chemin technique
 * brut (`guitare.electrique.lespaul`) à l'écran.
 */
export const formatClassificationLabel = (classification) => {
  if (!classification) return null;
  const { segments } = resolveClassification(classification);

  if (!segments) {
    // Valeur inconnue de la taxonomie : on affiche au moins son dernier segment, proprement.
    const fallback = String(classification).split('.').pop();
    return formatTaxonomyLabel(fallback);
  }

  const leaf = formatTaxonomyLabel(segments[segments.length - 1]);
  if (segments.length > 1 && segments[0] !== 'guitare') {
    return `${formatTaxonomyLabel(segments[0])} › ${leaf}`;
  }
  return leaf;
};

File: backend/taxonomy.py
"""Canonicalisation de la classification renvoyée par l'IA (2026-08-16).

Contexte : `analyzer.py` fait un `json.loads()` brut, sans `response_schema` ni énumération — rien
ne garantit que la valeur de `classification` appartienne réellement à la taxonomie. En pratique
l'IA renvoyait indifféremment un nom de feuille ("Parlor Standard"), un chemin complet
("guitare.electrique.lespaul") ou une valeur inventée, le prompt ne précisant pas le format attendu.

Le nom de feuille seul est **structurellement ambigu** : `Guitare Electrique` est à la fois une
feuille de `etui_housse.Etui_Rigide` (un étui) et le libellé de la branche `guitare.electrique`
(l'instrument). C'est ce qui faisait apparaître des étuis comme des guitares électriques.

Ce module ramène toute valeur à un **chemin complet en dot-notation**, seul format non ambigu, et
refuse explicitement de trancher quand le nom est porté par plusieurs branches — mieux vaut une
annonce non classée (corrigeable à la main depuis l'app) qu'une annonce rangée dans la mauvaise
catégorie sans que personne ne s'en aperçoive.
"""

import unicodedata
import re

# Branche des contenants, dont CERTAINES feuilles portent des noms génériques d'instruments
# ("Guitare Acoustique", "Guitare Electrique") alors qu'elles désignent des étuis pour ces
# instruments. Un tel nom nu n'est jamais résolu : il est bien plus probable qu'il désigne
# l'instrument lui-même.
#
# ⚠️ Restreint le 2026-08-16 après l'audit en production : la règle bloquait AUPARAVANT toute
# feuille de cette branche, y compris `Housse_Souple`, `Gigbag Standard`, `ATA Road Case` ou
# `Etui_Rigide` — des termes d'étui purs, sans la moindre ambiguïté, que rien ne justifiait de
# refuser (≈19 annonces correctement classées rendues invisibles). Seuls sont désormais bloqués
# les noms qui COMMENCENT par un nom de famille d'instrument, déduit dynamiquement des racines
# de la taxonomie plutôt que codé en dur.
CONTAINER_BRANCHES = ('etui_housse',)

# Nombre minimum de segments d'un chemin partiel accepté pour une réparation par suffixe :
# en dessous, on retombe sur une résolution par nom nu, avec le risque d'ambiguïté que ça implique.
MIN_PARTIAL_PATH_SEGMENTS = 2


def normalize_segment(text):
    """Minuscules, sans accents, alphanumérique uniquement — pour UN segment de chemin."""
    if not text:
        return ''
    decomposed = unicodedata.normalize('NFD', str(text))
    without_accents = ''.join(c for c in decomposed if unicodedata.category(c) != 'Mn')
    return re.sub(r'[^a-z0-9]', '', without_accents.lower())


def normalize_path(text):
    """Normalise un chemin en CONSERVANT le séparateur '.'.

    Le point porte du sens : sans lui, "Guitare Electrique" et "guitare.electrique" produisent la
    même clé (le bug des étuis). Miroir exact de `src/utils/taxonomy.js::normalizePath` côté front.
    """
    segments = [normalize_segment(s) for s in str(text or '').split('.')]
    return '.'.join(s for s in segments if s)


def build_index(taxonomy):
    """Construit `(by_path, by_leaf)` à partir de l'arbre `taxonomy_master`.

    `by_leaf` associe un nom normalisé à la LISTE des nœuds qui le portent (pas à un seul) : c'est
    ce qui permet de détecter l'ambiguïté au lieu de l'écraser silencieusement.

    Un nom "porté" recouvre DEUX façons de désigner le même texte, et c'est le cœur du bug des
    étuis : `Guitare Electrique` est une feuille de `etui_housse.Etui_Rigide`, mais c'est aussi la
    lecture naturelle du CHEMIN `guitare.electrique` une fois ses points retirés. Les deux sont
    donc enregistrés sous la même clé, ce qui rend le nom nu détectable comme ambigu — quel que
    soit le sens dans lequel l'IA l'a employé. Sans ça on ne fait que déplacer l'erreur : avant le
    correctif un étui devenait une guitare, avec un index par feuille seule une guitare deviendrait
    un étui.
    """
    by_path = {}
    by_leaf = {}

    def register(segments):
        by_path[normalize_path('.'.join(segments))] = list(segments)
        for key in {normalize_segment(segments[-1]), normalize_path('.'.join(segments)).replace('.', '')}:
            if key:
                by_leaf.setdefault(key, []).append(list(segments))

    def traverse(node, current):
        if isinstance(node, list):
            for item in node:
                register(current + [item])
        elif isinstance(node, dict):
            for key, value in node.items():
                segments = current + [key]
                register(segments)
                traverse(value, segments)

    traverse(taxonomy or {}, [])
    return by_path, by_leaf


def _instrument_roots(by_path):
    """Racines de la taxonomie désignant des objets (guitare, amplificateur…), hors contenants."""
    roots = {segments[0] for segments in by_path.values()}
    return {normalize_segment(r) for r in roots if r not in CONTAINER_BRANCHES}


def _repair_partial_path(classification, by_path):
    """Étend un chemin INCOMPLET vers l'unique chemin canonique qui s'y termine.

    L'audit en production a montré que l'IA produit régulièrement des chemins presque bons :
    racine oubliée (`electrique.solid_body.Double_Cut.SG`), niveau intermédiaire sauté
    (`guitare.acoustique_acier.Travel.Baby / Mini`, sans `specialites`) ou racine renommée depuis
    (`accessoire_etui.protection.Etui_Rigide.…`). Tous désignent sans ambiguïté une seule
    catégorie — les jeter revenait à perdre de l'information déjà correcte.

    On cherche donc les chemins canoniques qui SE TERMINENT par les segments fournis, en retirant
    au besoin les premiers segments de l'entrée (racine inventée). Un résultat n'est accepté que
    s'il est UNIQUE et qu'il reste au moins `MIN_PARTIAL_PATH_SEGMENTS` segments à comparer.
    """
    segments = [s for s in normalize_path(classification).split('.') if s]
    canonical_paths = list(by_path.values())

    while len(segments) >= MIN_PARTIAL_PATH_SEGMENTS:
        suffix = '.' + '.'.join(segments)
        matches = {
            '.'.join(path) for path in canonical_paths
            if normalize_path('.'.join(path)).endswith(suffix)
        }
        if len(matches) == 1:
            return matches.pop()
        if len(matches) > 1:
            return None  # Plusieurs cibles possibles : on ne devine pas.
        segments = segments[1:]  # Racine inventée/obsolète : on la retire et on réessaie.

    return None


def canonicalize(classification, taxonomy, index=None):
    """Ramène une valeur de classification à un chemin canonique.

    Retourne `(chemin_ou_None, raison)` où `raison` vaut :
      - 'exact_path'   : déjà un chemin complet valide
      - 'leaf'         : nom de feuille unique, étendu en chemin complet
      - 'partial_path' : chemin incomplet étendu vers l'unique chemin canonique correspondant
      - 'ambiguous'    : nom porté par plusieurs branches → NON résolu volontairement
      - 'unknown'      : absent de la taxonomie
      - 'empty'        : rien à résoudre
    """
    if not classification:
        return None, 'empty'

    by_path, by_leaf = index if index else build_index(taxonomy)

    path_key = normalize_path(classification)
    if path_key in by_path:
        return '.'.join(by_path[path_key]), 'exact_path'

    # Nom nu : on dédoublonne les chemins candidats (un même nœud peut être enregistré sous
    # plusieurs clés). Plus d'un nœud distinct = ambigu, on refuse de trancher.
    candidates = by_leaf.get(normalize_segment(classification)) or []
    distinct = sorted({'.'.join(c) for c in candidates})
    if len(distinct) == 1:
        segments = distinct[0].split('.')
        # Feuille d'un contenant : refusée seulement si son nom commence par un nom de famille
        # d'instrument ("Guitare Acoustique" → probablement l'instrument, pas l'étui). Les termes
        # d'étui purs ("Housse_Souple", "ATA Road Case") passent normalement.
        if segments[0] in CONTAINER_BRANCHES:
            leaf_key = normalize_segment(segments[-1])
            if any(leaf_key.startswith(root) for root in _instrument_roots(by_path)):
                return None, 'ambiguous'
        return distinct[0], 'leaf'
    if len(distinct) > 1:
        return None, 'ambiguous'

    # Chemin incomplet (racine oubliée, niveau sauté, racine obsolète) : extensible sans ambiguïté.
    repaired = _repair_partial_path(classification, by_path)
    if repaired:
        return repaired, 'partial_path'

    return None, 'unknown'
